src/js/bs3/module/HintPopover.js

```diff
@@ -0,0 +1,273 @@
+  'summernote/base/core/range',
+  'summernote/base/core/key'
+], function (func, list, dom, range, key) {
+  var HintPopover = function (summernote) {
+    var self = this;
+    var ui = $.summernote.ui;
+
+    var $note = summernote.layoutInfo.note;
+    var hint = summernote.options.hint || false;
+
+    if (hint && !(hint instanceof Array)) {
+      hint = [hint];
+    }
+
+    var KEY = key.code;
+
+    var DROPDOWN_KEYCODES = [KEY.UP, KEY.DOWN, KEY.ENTER];
+
+    var $popover = ui.popover({
+      className: 'note-hint-popover',
+      callback : function ($node) {
+        $node.css({
+          'min-width': '100px',
+          'padding': '2px'
+        });
+      }
+    }).render().appendTo('body');
+
+    var $popoverContent = $popover.find('.popover-content');
+
+
+    this.timer = null;
+
+
+    this.events = {
+      'summernote.keyup': function (e, nativeEvent) {
+        self.update(nativeEvent);
+      },
+      'summernote.keydown' : function (e, nativeEvent) {
+        self.updateKeydown(nativeEvent);
+      }
+    };
+
+    this.initialize = function () {
+      if (!hint) {
+        return;
+      }
+      dom.attachEvents($note, this.events);
+
+      $popoverContent.on('click', '.hint-item', function () {
+        $popoverContent.find('.active').removeClass('active');
+        $(this).addClass('active');
+        self.replace();
+        self.hide();
+      });
+    };
+
+    this.destroy = function () {
+      if (!hint) {
+        return;
+      }
+      dom.detachEvents($note, this.events);
+      $popoverContent.off('click');
+    };
+
+    this.activate = function (item) {
+      var $activeItem = $(item);
+      $popoverContent.find('.active').removeClass('active');
+      $activeItem.addClass('active');
+
+      this.scrollTo($activeItem);
+    };
+
+    this.scrollTo = function ($node) {
+      var $parent = $node.parent().parent();
+      $parent[0].scrollTop = $node[0].offsetTop - ($parent.innerHeight() / 2);
+    };
+
+    this.moveDown = function () {
+      var $old = $popoverContent.find('.hint-item.active');
+      var $next = $old.next();
+
+      if ($next.length) {
+        this.activate($next);
+      } else {
+        var $parentNext = $old.parent().next();
+
+        if (!$parentNext.length) {
+          $parentNext = $popoverContent.find('.hint-group').first();
+        }
+
+        this.activate($($parentNext).find('.hint-item').first());
+      }
+    };
+
+    this.moveUp = function () {
+      var $old = $popoverContent.find('.hint-item.active');
+      var $prev = $old.prev();
+
+      if ($prev.length) {
+        this.activate($prev);
+      } else {
+        var $parentPrev = $old.parent().prev();
+
+        if (!$parentPrev.length) {
+          $parentPrev = $popoverContent.find('.hint-group').last();
+        }
+
+        this.activate($($parentPrev).find('.hint-item').last());
+      }
+    };
+
+    this.replace = function () {
+      var wordRange = $popover.data('wordRange');
```

Comment:
You don't have to save status with data. How about save it this scope? (eg: this.wordRange)